File: Languages/C/Day026.c
// Languages - C - Day 26
// Foundational Concepts: Basics, syntax, loops, conditionals, functions, and variables.

// Create a function that determines if a given number is a Happy Number.

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

// Function to calculate the sum of squares of digits
int sumOfSquares(int num)
{
    int sum = 0;
    while (num > 0)
    {
        int digit = num % 10;
        sum += digit * digit;
        num /= 10;
    }
    return sum;
}

// Function to determine if a number is a Happy Number
bool isHappyNumber(int num)
{
    int seen[1000] = {0};
    while (num != 1 && seen[num] == 0)
    {
        seen[num] = 1;
        num = sumOfSquares(num);
    }
    return num == 1;
}

int main()
{
    int num;
    char choice;

    do
    {
        printf("Enter a number: ");
        while (scanf("%d", &num) != 1)
        {
            while (getchar() != '\n')
                ;
            printf("Invalid input. Please enter an integer: ");
        }

        if (num < 0)
        {
            printf("Negative numbers cannot be Happy Numbers.\n");
        }
        else
        {
            if (isHappyNumber(num))
            {
                printf("%d is a Happy Number.\n", num);
            }
            else
            {
                printf("%d is not a Happy Number.\n", num);
            }
        }

        printf("Do you want to check another number? (y/n): ");
        while (getchar() != '\n')
            ;
        choice = getchar();

    } while (choice == 'y' || choice == 'Y');

    printf("Exiting program. Goodbye!\n");

    return 0;
}

File: Languages/TypeScript/Day026.ts
// Languages - TypeScript - Day 26
// Foundational Concepts: Basics, syntax, loops, conditionals, functions, and variables.

// Create a function that determines if a given number is a Happy Number. 


function isHappyNum(num: number): boolean {
    const seen = new Set<number>();
    while (num !== 1 && !seen.has(num)) {
        seen.add(num);
        num = sumOfSquare(num);
    }
    return num === 1;
}

function sumOfSquare(num: number): number {
    return num.toString().split('').reduce((sum, digit) => sum + Math.pow(parseInt(digit), 2), 0);
}

// Use Node.js input handling
const readline = require('readline').createInterface({
    input: process.stdin,
    output: process.stdout
});

readline.question("Enter a number: ", (userInput: string) => {
    const inputNum = parseInt(userInput, 10);
    if (isNaN(inputNum)) {
        console.log("Invalid input. Please enter an integer.");
    } else if (inputNum < 0) {
        console.log("Negative numbers are not considered Happy Numbers.");
    } else {
        const result = isHappyNum(inputNum);
        console.log(`The number ${inputNum} is ${result ? "a Happy Number" : "not a Happy Number"}.`);
    }
    readline.close();
});
